alibi/explainers/counterfactual_tf2.py

```diff
@@ -0,0 +1,885 @@
+from functools import partial, singledispatch
+from tensorflow.keras.optimizers import Adam
+from tensorflow.keras.optimizers.schedules import PolynomialDecay
+from typing import Callable, Dict, List, Optional, Tuple, Union, TYPE_CHECKING
+
+
+if TYPE_CHECKING:  # pragma: no cover
+    import keras  # noqa
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: ALEX: TBD: Move the casting generic function to an appropriate utility file.
+# TODO: ALEX: This is a simple draft to get the CF code to work. More generally, we would
+#  define an object/series of functions to check if PyTorch/TF/both are installed and return an
+#  appropriate casting function.
+
+@singledispatch
+def to_numpy_arr(X: Union[tf.Tensor, tf.Variable, np.ndarray]):
+
+    supported_types = [tf.Tensor, tf.Variable, np.ndarray]
+    raise TypeError(f"Expected input object type to be one of {supported_types} but got type {type(X)}!")
+
+
+@to_numpy_arr.register
+def _(X: np.ndarray) -> np.ndarray:
+    return X
+
+
+@to_numpy_arr.register
+def _(X: tf.Tensor) -> np.ndarray:
+    return X.numpy()
+
+
+@to_numpy_arr.register
+def _(X: tf.Variable) -> np.ndarray:
+    return X.numpy()
+
+
+# TODO: ALEX: TBD I think it's best practice to define all the custom
+#  errors in one file (e.g., exceptions.py/error.py). We
+#  could standardise this across the library
+
+
+class CounterfactualError(Exception):
+    pass
+
+
+CF_SUPPORTED_DISTANCE_FUNCTIONS = ['l1']
+
+
+def counterfactual_loss(instance: tf.Tensor,
+                        cf: tf.Variable,
+                        lam: float,
+                        feature_scale: tf.Tensor,
+                        pred_probas: Optional[tf.Variable] = None,
+                        target_probas: Optional[tf.Tensor] = None,
+                        distance_fcn: str ='l1',
+                        ) -> tf.Tensor:
+
+    # TODO: ALEX: TBD: TO SUPPORT BLACK-BOX, PYTORCH WE NEED A SIMILAR FUNCTION AND A HIGHER LEVEL ROUTINE TO
+    #  "DISPATCH" AMONG THESE FUNCTIONS
+
+    # TODO: ALEX: DOCSTRING
+
+    distance = lam * distance_loss(instance, cf, feature_scale, distance_fcn=distance_fcn)
+    if pred_probas is None or target_probas is None:
+        return distance
+    pred = pred_loss(pred_probas, target_probas)
+    return distance + pred
+
+
+def distance_loss(instance: tf.Tensor,
+                  cf: tf.Variable,
+                  feature_scale: tf.Tensor,
+                  distance_fcn: str = 'l1') -> tf.Tensor:
+
+    # TODO: ALEX: DOCSTRING
+
+    if distance_fcn not in CF_SUPPORTED_DISTANCE_FUNCTIONS:
+        raise NotImplementedError(f"Distance function {distance_fcn} is not supported!")
+
+    ax_sum = tuple(np.arange(1, len(instance.shape)))
+
+    return tf.reduce_sum(tf.abs(cf - instance) / feature_scale, axis=ax_sum, name='l1')
+
+
+def pred_loss(pred_probas: tf.Variable, target_probas: tf.Tensor) -> tf.Tensor:
+
+    # TODO: ALEX: DOCSTRING
+
+    return tf.square(pred_probas - target_probas)
+
```

Comment:
I think that often improvements to the counterfactual method for specific problems will be made by modifying the loss functions (distance/pred) slightly to be more suitable for the problem at hand. Same goes for improvements of the CF method itself. 2 key parts are the optimisation function and process. Changes to the optimisation process (e.g. using a genetic algorithm) would usually involve bigger code changes anyway, but changes to the objective function should be very straightforward. Allowing users to pass custom loss functions for the distance and pred loss terms would be very convenient. These functions might need a specific signature but I think that's fine as users can always pass partially parameterised functions.